src/adversarial/neurify.jl

```diff
@@ -126,27 +162,39 @@ function constraint_refinement(solver::Neurify, nnet::Network, reach::SymbolicIn
     return intervals
 end
 
-function get_nodewise_gradient(nnet::Network, LΛ::Vector{Vector{Float64}}, UΛ::Vector{Vector{Float64}})
+function get_nodewise_gradient(solver::Neurify, nnet::Network, LΛ::Vector{Vector{Float64}}, UΛ::Vector{Vector{Float64}})
     n_output = size(nnet.layers[end].weights, 1)
     n_length = length(nnet.layers)
-    LG = Matrix(1.0I, n_output, n_output)
-    UG = Matrix(1.0I, n_output, n_output)
+    LG = ones(n_output)
+    UG = ones(n_output)
     max_tuple = (0, 0, 0.0)
     for (k, layer) in enumerate(reverse(nnet.layers))
         i = n_length - k + 1
-        for j in size(layer.bias)
-            if LΛ[i][j] ∈ (0.0, 1.0) && UΛ[i][j] ∈ (0.0, 1.0)
-                max_gradient = max(abs(LG[j]), abs(UG[j]))
-                if max_gradient > max_tuple[3]
-                    max_tuple = (i, j, max_gradient)
+        if layer.activation != Id() 
+            # Not sure whether this is right, but by experiments, this caused infinity loop. 
+            # Because a split didn't reduce over-approximation at all (the input set is not bisected).
+            for j in 1:size(layer.bias,1)
+                if in((i,j), solver.splitted)
+                    # To prevent infinity loop
+                    # Becuase the over-approximation, a split may not bisect the input set.
+                    # But in some cases (which I don't have an example, just a sense)
+                    # a node is indeed can be splitted twice.
+                    continue
+                end
+                if (0 < LΛ[i][j] < 1) && (0 < UΛ[i][j] < 1)
+                    max_gradient = max(abs(LG[j]), abs(UG[j]))
+                    if max_gradient >= max_tuple[3]
+                        max_tuple = (i, j, max_gradient)
+                    end
                 end
             end
         end
         i >= 1 || break
-        LG_hat = max.(LG, 0.0) * Diagonal(LΛ[i]) + min.(LG, 0.0) * Diagonal(UΛ[i])
-        UG_hat = min.(UG, 0.0) * Diagonal(LΛ[i]) + max.(UG, 0.0) * Diagonal(UΛ[i])
-        LG, UG = interval_map_right(layer.weights, LG_hat, UG_hat)
+        LG_hat = Diagonal(LΛ[i]) * max.(LG, 0.0) + Diagonal(UΛ[i]) * min.(LG, 0.0)
+        UG_hat = Diagonal(LΛ[i]) * min.(UG, 0.0) + Diagonal(UΛ[i]) * max.(UG, 0.0)
+        LG, UG = interval_map(copy(transpose(layer.weights)), LG_hat, UG_hat)
```

Comment:
I'm guessing you added the copy to avoid an error associated with `Adjoint{...}`? If so, the right thing to do is actually to change the type annotation in the definition of interval_map (see below) to `AbstractMatrix`
https://github.com/sisl/NeuralVerification.jl/blob/b99f427a77538f0863899b9c2f5d2ff68856759f/src/utils/util.jl#L256

Reply:
Will do.